corehq/celery_monitoring/heartbeat.py

```diff
@@ -0,0 +1,67 @@
+from __future__ import absolute_import
+from __future__ import print_function
+
+import datetime
+
+from celery.task import periodic_task
+
+from corehq.util.datadog.gauges import datadog_gauge
+from corehq.util.quickcache import quickcache
+
+
+HEARTBEAT_FREQUENCY = datetime.timedelta(seconds=10)
+HEARTBEAT_CACHE_TIMEOUT = datetime.timedelta(days=2)
+
+
+class HeartbeatNeverRecorded(Exception):
+    pass
+
+
+class Heartbeat(object):
+    def __init__(self, queue):
+        self.queue = queue
+
+    @quickcache(['self.queue'], timeout=HEARTBEAT_CACHE_TIMEOUT.total_seconds())
+    def get_last_seen(self):
+        # This function relies on the cache getting set manually in mark_seen
```

Comment:
I think this would be clearer with a django model instead of relying on quickcache. It'll be at most 20(?) records so it'll likely be cached in memory anyways.

Could also add a django admin model if desired

Reply:
That's interesting. This felt simpler to me for some reason. Is there a particular failure mode you're thinking of with keeping it in redis, or a particular practical advantage for keeping it in postgres? Or were you more thinking about just how this looks in code? For now I'm inclined to keep it the way it is, but I'm interested in hearing more

Reply:
A) I don't think we should use redis as a primary data store especially considering [cache eviction we've seen on prod](https://dimagi-dev.atlassian.net/browse/PLAT-136). (though debatable if redis vs datadog is our primary for this) 

B) Using a django model doesn't require explicit exception handling (could use get_or_create and create with a null last seen)

C) is datadog down or experiencing slow ingest rates? look at a django admin page for latest status

D) a django model more clearly documents the data we are capturing. Right now I think its kind of unclear what we're tracking

> This felt simpler to me for some reason

I think this is simpler if you are familiar with quickcache and can quickly get up to speed on this module (the comments help a lot). A django model may be more approachable to someone who doesn't have as good of an understanding of our custom caching. 

It also doesn't fit in with my mental model of what a "cache" is or is supposed to do so it feels like we're making quickcache much more generic than a "cache"

Reply:
I'm going to roll this out as is, using redis as the backend here. I'm onboard that the code around this is confusing for someone who's not familiar with the internals of quickcache, and isn't a normal use-case for quickcache, so it'd probably be cleaner to use actual cache.set and cache.get statements.

I don't think I share the view that storing in postgres is meaningfully better. For comparison, the current heartbeat check does almost exactly what I'm doing here, storing in redis, and I don't think we've ever had an issue with that getting evicted.

That said, I'm open to revisiting and revising. I mostly would like to get a V1 of this out today so we can see how it works as a celery metric—we currently don't have a meaningful celery check.

Reply:
And I do appreciate the reasoning and thoughtful review. (I probably should have led with that.) I'm going to make the changes to make the redis code more conventional and readable, and I'm sure I will refer back to the details in your last comment when iterating on this

Reply:
To be fair, my main issue with using redis as the storage medium is mostly idealistic/expectation setting. Its way easier to be able to answer the question "what data do we collect and store on HQ?" by saying "look at postgres/models.py" as that sets an obvious place to look for the data we are tracking when approaching a new app. I don't think postgres is meaningfully better than redis here either. I do think that using django's API and conventions is meaningfully better than rolling our own or depending on some of our own custom code.

Anyways I think you addressed the main concern here as the code is clearer thanks to using cache directly instead of relying on quickcache's internals. And at least there's a class that defines the interface to find heartbeats.

My nits aren't really enough to block this PR. Copule of thoughts around this(wouldn't be an issue with django model):

1. Currently this manually sets a redis key `heartbeat:{queue_name}`. I don't foresee name conflicts coming up soon, but how do we define what these keys should be? Do we have an exhaustive list of the namespaces that are currently taken?
2. Would it make sense to have `HeartbeatCache` in the models.py file since its essentially a database model?